src/libraries/Common/tests/System/Net/WebSockets/WebSocketStream.cs

```diff
@@ -0,0 +1,301 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Threading.Tasks;
+using System.IO;
+using System.Threading;
+
+namespace System.Net.WebSockets
+{
+    public class WebSocketStream : Stream
+    {
+        // Used by the class to hold the underlying socket the stream uses.
+        private readonly WebSocket _streamSocket;
+
+        // Whether the stream should dispose of the socket when the stream is disposed
+        private readonly bool _ownsSocket;
+
+        // Used by the class to indicate that the stream is m_Readable.
+        private bool _readable;
+
+        // Used by the class to indicate that the stream is writable.
+        private bool _writeable;
+
+        // Whether Dispose has been called. 0 == false, 1 == true
+        private int _disposed;
+
+        public WebSocketStream(WebSocket socket)
+            : this(socket, FileAccess.ReadWrite, ownsSocket: false)
+        {
+        }
+
+        public WebSocketStream(WebSocket socket, bool ownsSocket)
+            : this(socket, FileAccess.ReadWrite, ownsSocket)
+        {
+        }
+
+        public WebSocketStream(WebSocket socket, FileAccess access)
+            : this(socket, access, ownsSocket: false)
+        {
+        }
+
+        public WebSocketStream(WebSocket socket, FileAccess access, bool ownsSocket)
+        {
+            if (socket == null)
+            {
+                throw new ArgumentNullException(nameof(socket));
+            }
+            if (socket.State != WebSocketState.Open)
+            {
+                throw new IOException("The operation is not allowed on non-connected sockets.");
+            }
+
+            _streamSocket = socket;
+            _ownsSocket = ownsSocket;
+
+            switch (access)
+            {
+                case FileAccess.Read:
+                    _readable = true;
+                    break;
+                case FileAccess.Write:
+                    _writeable = true;
+                    break;
+                case FileAccess.ReadWrite:
+                default: // assume FileAccess.ReadWrite
+                    _readable = true;
+                    _writeable = true;
+                    break;
+            }
+        }
+
+        public WebSocket Socket => _streamSocket;
+
+        protected bool Readable
+        {
+            get { return _readable; }
+            set { _readable = value; }
+        }
+
+        protected bool Writeable
+        {
+            get { return _writeable; }
+            set { _writeable = value; }
+        }
+
+        public override bool CanRead => _readable;
+
+
+        public override bool CanSeek => false;
+
+
+        public override bool CanWrite => _writeable;
+
+
+        public override bool CanTimeout => true;
+
+
+        public override long Length
+        {
```

Comment:
` public override long Length => throw new NotSupportedException("This stream does not support seek operations.");`